Работа с пользователями и фильмами

YetiSPb opened the pull request:
Созданы объекты
- Film
- User

Созданы контроллеры для работы с ними
- FilmController
- UserController

Созданы тесты для контроллеров
- FilmControllerTest
- UserControllerTest

Took 2 minutes

YetiSPb committed:
Работа с пользователями и фильмами

Созданы объекты
- Film
- User

Созданы контроллеры для работы с ними
- FilmController
- UserController

Созданы тесты для контроллеров
- FilmControllerTest
- UserControllerTest

Took 2 minutes

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,85 @@
+    private int countId = 0;
+
+    @PostMapping()
+    public ResponseEntity<Film> addFilm(@RequestBody @NotNull Film film) {
```
Пока можно так - но в следующей работе потребуется добавить глобальный обработчик ошибок, и тогда очень рекомендую перейти здесь на возвращение просто Film - код станет проще и понятнее. Аналогично во всех методах контроллерах

annaalkh replied:
Не уверена, как здесь задумывалось, но использование аннотации org.jetbrains.annotations.NotNull может ввести в заблуждение - выглядит так, как будто для переданных пользователем данных есть проверка, но на самом деле эта аннотация используется только IntellijIDEA для статического анализа кода

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,85 @@
+    public ResponseEntity<Film> updateFilm(@RequestBody @NotNull Film film) throws ValidationException {
+
+        if (films.get(film.getId()) == null) {
+            throw new ValidationException("Фильм не найден!");
```
Здесь нелогично - почему это ValidationException никем не обрабатывается, а в следующем блоке такое же исключение - обрабатывается? Должно быть единообразие - я бы рекомендовала не обрабатывать ValidationException нигде. Пусть пользователь получит свою ошибку, а в следующей работе сделаем ответ чуть более информативным

YetiSPb replied:
Тутт не понял. Если убрать try catch то как я определю какой ответ отправлять клиенту?

annaalkh replied:
Тут как раз помог бы глобальный обработчик ошибок для ValidationException, он бы и определял ответ пользователю (но убрать все под try - тоже вариант)

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,85 @@
+
+        try {
+            generalValidateFilm(film);
+            films.remove(film.getId());
```
А зачем эта строчка? put все равно заменит объект

YetiSPb replied:
Для надёжности ))). 
Просто забыл убрать..

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,85 @@
+
+    private void generalValidateFilm(@NotNull Film film) throws ValidationException {
+
+        if (film.getName() == null || film.getName().equals(""))
```
Лучше проверять, что строка пустая, методом film.getName().isEmpty. А на практике для подобных проверок обычно используют метод isEmpty из библиотеки apache commons (или аналогичной - https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html), тогда можно будет убрать и проверку на null

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,90 @@
+@Slf4j
+@RestController
+@RequestMapping("/users")
+public class UserController {
```
Комментарии аналогичны FilmController

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,90 @@
+
+    private void generalValidateUser(@NotNull User user) throws ValidationException {
+
+        Pattern patternEMail = Pattern.compile("^[a-zA-Z0-9_+&*-]+(?:\\." +
```
На практике для проверок email очень рекомендуется использовать готовую аннотацию EMail - это гораздо надежнее, чем вручную составлять регулярное выражение

YetiSPb replied:
Т.е. все проверки можно было сделать в классах User и Film.
Смотрел эту статью https://struchkov.dev/blog/ru/spring-boot-validation/

annaalkh replied:
Да, все верно - это обычно самый удобный вариант

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,90 @@
+            throw new ValidationException("Электронная почта не может быть пустой и должна содержать символ @");
+        }
+
+        if (user.getLogin().equals("") || user.getLogin().indexOf(" ") > 0 || user.getLogin() == null)
```
Первым должно быть условие проверки на null, иначе возможно NPE

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,90 @@
+            user.setName(user.getLogin());
+        }
+
+        if (user.getBirthday().isAfter(LocalDate.now())) {
```
А если дата совсем не указана? Тогда этот код упадет

YetiSPb replied:
В Творческом задании не указано, думал что проверка будет не на моей стороне. Но на всякий случай добавил

Для User:
электронная почта не может быть пустой и должна содержать символ @;
логин не может быть пустым и содержать пробелы;
имя для отображения может быть пустым — в таком случае будет использован логин;
дата рождения не может быть в будущем.

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/exception/ValidationException.java:
```diff
@@ -0,0 +1,12 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
```
В исключениях лучше логгер не использовать - исключением могут пользоваться разные классы со своими предпочтениями в логировании, это их ответственность - что и когда логировать

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/controller/exception/ValidationException.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ValidationException extends Exception {
```
Лучше сделать исключение непроверяемым, чтобы не приходилось его всегда отлавливать

annaalkh reviewed (commented):
Евгений, привет! В целом все отлично, есть отдельные нюансы которые можно доработать - отметила в комментариях

YetiSPb committed:
Работа с пользователями и фильмами

Исправлена работа в контроллерах
- FilmController
- UserController

Took 1 hour 49 minutes

YetiSPb commented:
Поправил все комменты. За советы спасибо.

annaalkh reviewed (approved):
Теперь отлично!
